### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by valeryan**

### CODE REVIEW

**Comment by jonathanbossenger** [comment]
@valeryan thanks for opening this PR. 

I was able to determine why my local configuration was broken, so my plan this week is to test this PR and make sure everything still works after the dependencies update. 

I will report back once my testing is done.

**Comment by jonathanbossenger** [comment]
@valeryan quick question on testing.

When running the extension in debug/development mode, does the Extension Development Host version automatically use the code from the local directory, and not the installed version of the extension, or do I need to uninstall the installed version first?

I want to be able to do things like [delete the executables via the GUI to test this bug](https://github.com/valeryan/vscode-phpsab/issues/90).

**Comment by valeryan** [comment]
Yes it does replace the installed version. You can then load any code test.

**Comment by jonathanbossenger** [comment]
Perfect, thank you.

**Comment by jonathanbossenger** [comment]
@valeryan I have tested this, and I think it's ready to merge.

1. I tested with my typical setup, which uses phpcs and phpcbf configured in the plugin settings
2. I tested in a standalone app, which has its own versions of phpcs and phpcbf, with a custom phpcs.xml configuration
3. I tested the fix for #90 and it worked as expected when deleting the executable paths in the GUI

Finally, I updated the readme to include a link to the Docker fork, to resolve #88

**Comment by valeryan** [comment]
Also if it's helpful to you. I have this https://github.com/valeryan/vscode-phpcs-wp-plugin-example that I use for a test bed.

**Comment by valeryan** [comment]
@jonathanbossenger let me know when you are full satisfied with this and I will merge it to master and draft a release.


Thanks

**Comment by jonathanbossenger** [comment]
Will do.

I'm traveling to the US tomorrow, so I'll have time to test with your test bed on the flight. Will confirm once I've done that.

**Comment by jonathanbossenger** [comment]
@valeryan checking in here. 

The morning of my trip my first flight was canceled, and I spent the rest of the morning rebooking the flight, and then frantically packing to leave 2 hours earlier than planned. Suffice it to say I had no time to test on the trip.

I'll be picking this up next week, and I will report back once tested.

I see there is also a package-lock.json conflict, which I'll be sure to resolve before the merge.

**Comment by macr1408** [comment]
@jonathanbossenger good job here, from what I understood only testing is pending right? I would love to see this PR being merged so if I can help in any way to speed up the process feel free to tag me.

**Comment by jonathanbossenger** [comment]
Hey @macr1408 yes, mostly it's just testing, I did some initial testing myself, but since a few weeks ago I've been struggling to recover from a cold, and have not had much time to test. If you're able to test and verify things all work as expected, that would be helpful.

**Comment by macr1408** [comment]
Sorry for the late reply, I've been really busy lately too. 

Ok so first, I was expecting to test the docker features because that's what I use most at my work, but looks like this PR only adds a comment to the README and focuses on multi-root workspace, which is fine but sadly I won't be really helpful here because I don't use that feature (sorry).

Anyways I still tried to help and downloaded this https://github.com/valeryan/multi-root-example, and it's not working at least for me, it does not find executables in directory `root1` or `root2` (however if I open those directories individually they work perfectly.) But don't take this as a reliable result, it might be just me not configuring this thing properly.

Good luck :-)

**Comment by valeryan** [comment]
@jonathanbossenger, @seebeen, I have the repo set up so that one or both of you have to approve this branch in Github to release it. So I am waiting for you guys to feel like you have tested this branch and have at least enough confidence to say it's releasable and that we can reasonably handle any bugs that pop up.

I want to be clear as our working relationship on this project is very new. There are two of you now, which means you can even work together on a branch, approve, and release it without me. As long as a PR has at least one of us approving it, it can merge.

**Comment by seebeen** [comment]
Hallo. I'm still working on setting up automated unit tests for some stuff so I don't introduce regression.

My firsy order of business is improving the detection of the binary path and logging standardization.

After that i think moving the work to an lsp would be a good idea

File: src/configuration.ts [hunk 1]
```diff
@@ -13,164 +13,159 @@ import { PathResolver } from "./resolvers/path-resolver";
...
+  }
+
+  /**
+   * Get correct executable path from resolver
+   * @param settings
+   */
+  protected async resolveCBFExecutablePath(
+    settings: ResourceSettings
+  ): Promise<ResourceSettings> {
+    if (settings.executablePathCBF === null || settings.executablePathCBF === "") {
```
Comment (valeryan) [inline src/configuration.ts:102]: I think this can probably just be if (!settings.executablePathCBF)

File: src/configuration.ts [hunk 2]
```diff
@@ -13,164 +13,159 @@ import { PathResolver } from "./resolvers/path-resolver";
...
+  }
+
+  /**
+   * Get correct executable path from resolver
+   * @param settings
+   */
+  protected async resolveCSExecutablePath(
+    settings: ResourceSettings
+  ): Promise<ResourceSettings> {
+    if (settings.executablePathCS === null || settings.executablePathCS === "") {
```
Comment (valeryan) [inline src/configuration.ts:124]: I think this can probably just be if (!settings.executablePathCS)

**Comment by jonathanbossenger** [comment]
@valeryan thanks.

I will be honest, I was very keen to help out, and then life and other things got in the way, so I took a step back. But I am keen to get stuck in and help keep this extension updated. 

I'm going to be spending quite a bit of time using it to auto-format a bunch of code this week, so I will make sure to keep testing and report back.

@valeryan out of interest (I'm not an experienced VS Code extension developer) is there a way to create a 'beta' version of the extension from a specific branch, that I could install as my default VS Code extension for testing purposes, as opposed to running it in Debug mode in VS Code?

File: src/configuration.ts (hunk 1 above)
Comment (jonathanbossenger) [reply src/configuration.ts:102]: That's a good point, I will check.

File: src/configuration.ts (hunk 2 above)
Comment (jonathanbossenger) [reply src/configuration.ts:124]: That's a good point, I will check.

**Comment by valeryan** [comment]
@jonathanbossenger ahh Life, the thing that takes out most developers. 😜 About a "beta", you can package the extension and then install it to to replace the main version from marketplace. https://code.visualstudio.com/api/working-with-extensions/publishing-extension you can use vsce to package it up like its ready to publish and then install the generated package. I'll do a little write up on that and add it to the development.md doc.

**Comment by jonathanbossenger** [comment]
> I'll do a little write up on that and add it to the development.md doc.

Thank you @valeryan that would be greatly apprecaited.

**Comment by valeryan** [comment]
@seebeen I installed and tested the extension this afternoon. It worked well for me. I also ran the unit test you added for logging. We should be able to wire up a workflow that runs the unit test suite for PRs. It might be good to go ahead and merge and release this as 0.0.17 and then start working on your plans on a new dev branch for v1.0.0. What are your thoughts?

**Comment by valeryan** [comment]
@jonathanbossenger and @seebeen, I refactored the whole extension this weekend. I tried to modernize all the Typescript somewhat and remove the class constructs. I think all that's left to do on this pass to get it to a state that I feel merits the 1.0.0 status is to get a unit test built for the fixer and sniffer. I started writing unit tests for the path resolvers but I hate mocking stuff and sinon is new to me. All the path and fs stuff can be a bit painful to isolate from the unit test. If y'all are interested, then getting this test suite kitted out would be super cool.

**Comment by valeryan** [comment]
Also, I think might just merge this and start a new dev branch for building out the test. This PR is getting very huge on account of my fooling around. I am not going to draft a release yet but I normally try to keep my PRs way smaller.

**Comment by jonathanbossenger** [comment]
Thanks for all the updates @valeryan.

I am mostly in a JavaScript focused week or two at work (bad timing there) so I'm not going to be in a space to test the extension for PHP code soon. As soon as I'm back in PHP land, I'll be able to take a look.

**Pull request merged**